src/historical-stipend-calculator.ts

```diff
@@ -0,0 +1,342 @@
+  // First try to extract location after hyphen
+  const hyphenRegex = /- ([^,]+)$/;
+  const hyphenMatch = hyphenRegex.exec(eventName);
+  if (hyphenMatch) {
+    return hyphenMatch[1].trim();
+  }
+
+  // Then try to extract known city names from anywhere in the event name
+  const knownCities = {
+    Singapore: "Singapore",
+    Bangkok: "Bangkok",
+    Seoul: "Seoul",
+    Taipei: "Taipei",
+    "Hong Kong": "Hong Kong",
+    "San Francisco": "San Francisco",
+  };
+
+  for (const [cityName, normalizedName] of Object.entries(knownCities)) {
+    if (eventName.includes(cityName)) {
+      return normalizedName;
+    }
+  }
+
+  return "";
+}
+
+// Helper function to parse dates from the format "Month D–D, YYYY"
+function parseDates(dateString: string): { start: string; end: string } {
+  // Handle various date formats
+  const [datePart] = dateString.split(", ");
+
+  // Split the date part by the dash (either normal or en dash)
+  const [startPart, endPart] = datePart.split(/[–-]/);
+
+  const startMonth = startPart.trim().split(" ")[0];
+  const startDay = startPart.trim().split(" ")[1] || "1";
+
+  let endMonth, endDay;
+  if (endPart) {
+    endMonth = endPart.includes(" ") ? endPart.trim().split(" ")[0] : startMonth;
+    endDay = endPart.includes(" ") ? endPart.trim().split(" ")[1] : endPart.trim();
+  } else {
+    endMonth = startMonth;
+    endDay = startDay;
+  }
+
+  return {
+    start: `${startDay} ${startMonth}`,
+    end: `${endDay} ${endMonth}`,
+  };
+}
+
+// Function to calculate stipend with caching
+async function calculateStipend(record: HistoricalConference): Promise<StipendBreakdown> {
+  const destination = extractLocation(record["Event Name"]);
+  const dates = parseDates(record["Dates"]);
+
+  const cacheKey = createHashKey([
+    record["Event Name"],
+    destination,
+    dates.start,
+    dates.end,
+    ORIGIN,
+    COST_PER_KM,
+    BASE_LODGING_PER_NIGHT,
+    BASE_MEALS_PER_DAY,
+    BASE_LOCAL_TRANSPORT_PER_DAY,
+    BUSINESS_ENTERTAINMENT_PER_DAY,
+    record["Ticket Price (USD)"],
+    "historical_v5", // Increment version to invalidate cache
+  ]);
+
+  // Force recalculation for all conferences to use the new flight cost algorithm
+  // Skip cache completely
+
+  console.log(`Processing conference: ${record["Event Name"]} in ${destination}`);
+
+  // Calculate distance (in km) using the haversine formula with our city coordinates
+  const distanceKm = getDistanceKmFromCities(ORIGIN, destination, cityCoordinates);
+  console.log(`Distance from ${ORIGIN} to ${destination}: ${distanceKm.toFixed(1)} km`);
+
+  // Try to get flight cost from API first
+  const flightDates = generateFlightDates({
+    Conference: record["Event Name"],
+    Location: destination,
+    Category: "Historical",
+    Start: dates.start,
+    End: dates.end,
+    "Ticket Price": record["Ticket Price (USD)"],
+  });
+  const apiFlightPrice = await lookupFlightPrice(destination, flightDates);
```

Comment:
to avoid spamming errors and unnecessary computations check if the API key is set 
```suggestion
  // Try to get flight cost from API first
  let apiFlightPrice: number | null = null;
  if (process.env.SERPAPI_API_KEY) {
    apiFlightPrice = await lookupFlightPrice(destination, flightDates);
  } else {
    console.warn("SERPAPI_API_KEY is not set. Using the enhanced flight cost calculation model.");
  }
```